apache/metron
# METRON-1031: Management UI Cannot Start Topologies in Kerberized Environment

The Metron Management UI cannot start topologies in a kerberized environment.  This includes the Parser, Enrichment, and Indexing topologies.

It seems that Metron REST does not pass "-ksp" to the start topology commands. As a result, topologies that are started with the Management UI cannot start a producer against a kerberized Kafka.

### Changes

* Altered the `StormCLIWrapper` to add logging of the actual CLI commands that are executed.
* The `-ksp` argument is always used (with Kerberos or not) when launching the Parser, Enrichment, and Indexing topologies.
* Added a Spring property `kafka.security.protocol` that serves as the `-ksp` argument when launching the topologies.
* Ambari sets this Spring property based on Kafka's own configuration in Ambari.

### Testing

To manually test this, validate that all of the topologies can be stopped and started using the Management UI and Swagger UI both before and after kerberization.

1. Launch the Full Dev environment.
1. Stop then start the Parser topologies using the Management UI.
1. Ensure that data is being consumed by the topologies and landing in the Elasticsearch indices.
1. Kerberize the Full Dev environment [using these instructions](https://github.com/apache/metron/blob/master/metron-deployment/Kerberos-ambari-setup.md).
1. Kerberize the Sensor Stubs.

    Add the following to somewhere near the top of `/etc/init.d/sensor-stubs`.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    ```
    Add the `--security-protocol`.
    ```
    sed -i 's/node1:6667 --topic/node1:6667 --security-protocol PLAINTEXTSASL --topic/' /opt/sensor-stubs/bin/start-*-stub
    ```
    Restart the stubs.
    ```
    for sensorstub in bro snort; do
        service sensor-stubs stop ${sensorstub};
        service sensor-stubs start ${sensorstub};
    done
    ```
1. Ensure data is being sent to the Parsers from the Sensor Stubs.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic bro
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic snort
    ```
1. Ensure that the Parsers are sending data to the Enrichment topology.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic enrichments
    ```
1. Stop the Bro & Snort topology using the Management UI.
1. Start the Bro topology using the Management UI.
1. Ensure that the Bro Parser is sending data to the Enrichment topology.
    ```
    export KAFKA_OPTS="-Djava.security.auth.login.config=/usr/metron/0.4.1/client_jaas.conf"
    /usr/hdp/current/kafka-broker/bin/kafka-console-consumer.sh --zookeeper node1:2181 --security-protocol PLAINTEXTSASL --topic enrichments
    ```


### Pull Request Checklist
